### DISCUSSION TITLE
how to get request body(json) and response body?

### DISCUSSION THREAD

**Discussion by fenghaoyu** [fastapi/fastapi#7660] [Questions] [answered]
i'm botherd to find some solusion to record log for each request.
i need to record request params and response body etc.
i tried to use middleware like this.
```
    @app.middleware("http")
    async def add_process_time_header(request: Request, call_next):
        response = await call_next(request)

        # i got this right
        print(request.url, request.method)
        print(dict(request.headers.items()))
        print(dict(request.query_params.items()))
        print(response.status_code)
        print(request.url)

        # i can't get request body(json) and response body
        # print(json_param, response_body)

        return response
```
Thank you very much if there is any help.

**Comment by fenghaoyu** [comment]
i got a way to get request json body and response data.
```
    @app.middleware("http")
    async def add_process_time_header(request: Request, call_next):
        response = await call_next(request)

        # request json body
        json_param = await request.json()
        print(json_param)

        # response data, seems only useful when the response is default Response
        content = b''
        async for chunk in response.body_iterator:
            content += chunk
        response_data = str(content, encoding='utf-8')
        print(response_data )

        return response
```

**Comment by fenghaoyu** [comment]
> i got a way to get request json body and response data.
> 
> ```
>     @app.middleware("http")
>     async def add_process_time_header(request: Request, call_next):
>         response = await call_next(request)
> 
>         # request json body
>         json_param = await request.json()
>         print(json_param)
> 
>         # response data, seems only useful when the response is default Response
>         content = b''
>         async for chunk in response.body_iterator:
>             content += chunk
>         response_data = str(content, encoding='utf-8')
>         print(response_data )
> 
>         return response
> ```

That is unposible to solve the problem.
And i find this:
https://github.com/tiangolo/fastapi/issues/394
it's helpful.

**Comment by tiangolo** [comment, marked as answer]
That's probably a complex use case, but this might be useful too: https://fastapi.tiangolo.com/advanced/custom-request-and-route/

Thanks for reporting back and closing the issue.

**Comment by vjpr** [comment]
I have tried this and it results in:

```
raise RuntimeError("Response content shorter than Content-Length")
```

I guess because the iterator needs to be reset somehow.
